chunky/src/java/se/llbit/chunky/renderer/Renderer.java

```diff
@@ -1,101 +1,36 @@
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- * GNU General Public License for more details.
- * You should have received a copy of the GNU General Public License
- * along with Chunky.  If not, see <http://www.gnu.org/licenses/>.
- */
 package se.llbit.chunky.renderer;
 
-import se.llbit.chunky.renderer.scene.Scene;
-import se.llbit.chunky.resources.BitmapImage;
-import se.llbit.util.TaskTracker;
+import java.util.function.BooleanSupplier;
 
-import java.util.function.BiConsumer;
-import java.util.function.Consumer;
-
-/**
- * A renderer renders to a buffered image which is displayed by a render canvas.
- *
- * @author Jesper Öqvist <jesper@llbit.se>
- */
 public interface Renderer {
-  void setSceneProvider(SceneProvider sceneProvider);
-
-  void setCanvas(Repaintable canvas);
-
   /**
-   * Instructs the renderer to change its CPU load.
+   * Get a short id string representing this renderer.
    */
-  void setCPULoad(int loadPercent);
+  String getIdString();
 
   /**
-   * Set a listener for render completion.
+   * The post render callback. This should be run after rendering a frame.
+   * It will return {@code true} if the render should terminate.
    *
-   * @param listener a listener which is passed the total rendering
-   * time and average samples per second.
+   * Implementation details, this deals with:
+   *  * Checking if the render mode has changed
+   *  * Updating the task-tracker
+   *  * Repainting the canvas
+   *  * Updating the {@code bufferedScene}
    */
-  void setOnRenderCompleted(BiConsumer<Long, Integer> listener);
+  void setPostRender(BooleanSupplier callback);
 
   /**
-   * Set a listener for frame completion.
+   * This is called when a render is initiated.
    *
-   * @param listener a listener which is called when a frame completes
-   * with the current scene and the current samples per pixel.
+   * * It should render a frame, merge that frame with {@code manager.bufferedScene}, and update the spp values.
+   * * It should call the post-render callback (set in {@code setPostRender(callback)}) and terminate if it returns {@code true}.
    */
-  void setOnFrameCompleted(BiConsumer<Scene, Integer> listener);
-
-  void setSnapshotControl(SnapshotControl callback);
-
-  void setRenderTask(TaskTracker.Task task);
-
-  void addRenderListener(RenderStatusListener listener);
-
-  void removeRenderListener(RenderStatusListener listener);
-
-  void withBufferedImage(Consumer<BitmapImage> bitmap);
-
-  interface SampleBufferConsumer {
-    void accept(double[] samples, int width, int height);
-  }
-
-  void addSceneStatusListener(SceneStatusListener listener);
-
-  void removeSceneStatusListener(SceneStatusListener listener);
-
-  RenderStatus getRenderStatus();
-
-  /**
-   * Start up the renderer.
-   *
-   * <p>This should start all worker threads used by the renderer.
-   */
-  void start();
+  void render(InternalRenderManager manager) throws InterruptedException;
```

Comment:
is this _always_ an `InternalRenderManager`? What if a plugin replaced the `renderManagerFactory`?

Reply:
I was thinking that the `Renderer` interface is only for plugins that fit into the existing Chunky code well. If a plugin were to change so many things that it needs to override the `RenderManager` other renderers probably wouldn't work in the first place (or it wouldn't make sense to have other renderers).

Essentially, this allows simple renderer plugins to use specific things like the `RenderWorkerPool` or `bufferedScene` without forcing low level renderer plugins to provide it or implement "dummy" versions (if that makes any sense).

Reply:
This sounds reasonable 👍
Maybe call it `DefaultRenderManager` then?